### DISCUSSION TITLE
Potential Table Parsing Error RemarkGFM

### DISCUSSION THREAD

**Discussion by pjkaufman** [remarkjs/remark#1003] [Q&A] [answered]
I am using `remark-gfm` in order to help parse some markdown, but it seems that it is coming up with some results that I am not expecting.

Here is the markdown I am parsing:
``` plaintext
# Table 1
| Column 1 | Column 2 | Column 3 |
|----------|----------|----------|
| foo      | bar      | blob     |
| baz      | qux      | trust    |
| quux     | quuz     | glob     |
# Table 2
| Column 1 | Column 2 |
|----------|----------|
| foo      | bar      |
| baz      | qux      |
| quux     | quuz     |

New paragraph.
```

I am looking for just the tables present. I expect the results to include 2 tables, but for some reason I am only getting one table output.

The result I get for the table hits in terms of the actual text is:
``` plaintext
| Column 1 | Column 2 | Column 3 |
|----------|----------|----------|
| foo      | bar      | blob     |
| baz      | qux      | trust    |
| quux     | quuz     | glob     |
# Table 2
| Column 1 | Column 2 |
|----------|----------|
| foo      | bar      |
| baz      | qux      |
| quux     | quuz     |
```

I expect to get 2 tables:
``` plaintext
| Column 1 | Column 2 | Column 3 |
|----------|----------|----------|
| foo      | bar      | blob     |
| baz      | qux      | trust    |
| quux     | quuz     | glob     |
```
and
``` plaintext
| Column 1 | Column 2 |
|----------|----------|
| foo      | bar      |
| baz      | qux      |
| quux     | quuz     |
```

Please let me know if I am doing something wrong or if this is the intended functionality.

**Comment by ChristianMurphy** [comment, marked as answer]
@pjkaufman could you give a bit more context on what versions of remark and remark-gfm you're using, and how they're being used?

One of the easiest ways to communicate all that is to create a sandbox with an example of what you are trying.
Here are some starters:

| description | codesandbox | stackblitz |
| - | - | - |
| remark only (for markdown to markdown) | [codesandbox](https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/remark-with-parcel) | [stackblitz](https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/remark-with-parcel) |
| remark and rehype (for markdown to html) | [codesandbox](https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/remark-rehype-with-parcel) | [stackblitz](https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/remark-rehype-with-parcel) |
| react-markdown | [codesandbox](https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-create-react-app) | [stackblitz](https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-create-react-app) |

**Reply by pjkaufman** [reply]
@ChristianMurphy , here is the link to the sandbox in for using `remark-gfm`: https://codesandbox.io/s/relaxed-orla-9wkxgi?file=/src/index.js

The version of node I use is `v16.15.0`. The version of remark and remark-gfm:
``` JSON
"remark": "^14.0.1",
"remark-gfm": "^3.0.0", 
```

Check the console log output for what I am referring to. It may be I am doing something wrong.

**Reply by ChristianMurphy** [reply]
🤔 this may be related to https://github.com/micromark/micromark-extension-gfm-table/issues/4

**Reply by pjkaufman** [reply]
It seems they are, though that issue is listed as completed.

**Reply by wooorm** [reply]
Because that issue is completed. Christian is pointing to the fact that that fix might cause this problem.

**Reply by pjkaufman** [reply]
Gotcha. Does that mean I should see about getting that issue reoppened? As it seems to refer to what I am seeing here?

**Reply by ChristianMurphy** [reply]
I'm not sure it cause the issue, it may be another edge case related to it though.
I'd defer to @wooorm on preference for re-opening the old issue, or creating a new one.

**Reply by ChristianMurphy** [reply]
For posterity, this is now tracked in https://github.com/micromark/micromark-extension-gfm-table/issues/5

**Reply by pjkaufman** [reply]
I must not have been concise enough and would not like to bug wooorm any further, at least today, so I will just add this comment here. I have two examples on Code Sandbox that I intended to convey on the issue that was opened, but I did so poorly. Here is the issue I was seeing with the use of `remark-gfm` (note that all versions are now up to date, but same issue happens on the latest or a downgraded version).

Here is the rough code that creates the issue:
```Typescript
import { visit } from "unist-util-visit";
import { fromMarkdown } from "mdast-util-from-markdown";
import { gfm } from "micromark-extension-gfm";
import { gfmFromMarkdown } from "mdast-util-gfm";

const sourceMarkdown =
  "# Tables that are back to back with content in between\n\n| Column 1 | Column 2 | Column 3 |\n|----------|----------|----------|\n| foo      | bar      | blob     |\n| baz      | qux      | trust    |\n| quux     | quuz     | glob     |\n# Table 2 without Pipe at Start and End\n| Column 1 | Column 2 | Column 3 |\n| -------- | -------- | -------- |\nbar | baz | baq\nfoo | bar | bun\n";

const ast = fromMarkdown(sourceMarkdown, {
  extensions: [gfm()],
  mdastExtensions: [gfmFromMarkdown()]
});
console.log(ast);
let positions = [];
visit(ast, "table", (node) => {
  positions.push(node.position);
});

// Sort positions by start position in reverse order
positions.sort((a, b) => b.start.offset - a.start.offset);
for (const position of positions) {
  console.log(
    sourceMarkdown.substring(position.start.offset, position.end.offset)
  );
}
```

Here is the sandbox that imports only the necessary dependencies that are used in the above code: https://codesandbox.io/s/npm-playground-forked-dogubv?file=/src/index.js

The output looks great for the tables:
```markdown
| Column 1 | Column 2 | Column 3 |
| -------- | -------- | -------- |
bar | baz | baq
foo | bar | bun 
```
and
```markdown
| Column 1 | Column 2 | Column 3 |
|----------|----------|----------|
| foo      | bar      | blob     |
| baz      | qux      | trust    |
| quux     | quuz     | glob     | 
```

The Code Sandbox here adds `remark-gfm` v3.0.1 and it all of a sudden stops working the same: https://codesandbox.io/s/relaxed-orla-9wkxgi?file=/src/index.js

The output looks like so for markdown tables:
```markdown
| Column 1 | Column 2 | Column 3 |
|----------|----------|----------|
| foo      | bar      | blob     |
| baz      | qux      | trust    |
| quux     | quuz     | glob     |
# Table 2 without Pipe at Start and End
| Column 1 | Column 2 | Column 3 |
| -------- | -------- | -------- |
bar | baz | baq
foo | bar | bun 
```
I see the same issue locally even when removing the `package-lock.json`. Please feel free to let me know this is operator error and that it is not reproducible if that is the case. I just don't know how to move forward with getting this resolved if it is an error and not a caching issue as the associated issue was closed as un-reproducible.

I will let this sit and wait for a determination on user error or actual issue. Thanks!

**Reply by wooorm** [reply]
> I must not have been concise enough and would not like to bug wooorm any further

I appreciate that you take me locking the issue well. 👏

The reason for locking it is that I tried to reproduce your issues 3 times, and it isn’t reproducible. I mentioned this several times, and mentioned ways to solve it, but my solutions were ignored (or partially applied).

---

I believe you do not care about CodeSandbox, but care about your local project.
If you’d try to reproduce this locally, minimally, you’ll see that it isn’t reproducible.
Create a new folder, add each of [these files](https://github.com/micromark/micromark-extension-gfm-table/issues/5#issuecomment-1200209323), run `npm install`, and then run `node example.js`: it works.

> I see the same issue locally even when removing the `package-lock.json`.

Please try a new empty project, and, if you are using pnpm, clear its cache (PLEASE PERFORM THIS STEP).

If it works in a new project, but not in your old project, REMOVING `package-lock` IS NOT ENOUGH. Remove `node_modules`, run `npm install`, AND run `npm update`, etc.

If it still doesn’t work, please explain in-depth EXACTLY what tools you are using, what versions they have, and upload a zip of the tiniest possible reproduction.

**Reply by pjkaufman** [reply]
Thanks for pointing out the next steps as I am not the most familiar with `npm` and `node` when it comes to caching and how to isolate issue around versions potentially having issues (it may just be operator error in this case). The example provided seems to skip using the AST generated by `remark-gfm` or its dependencies and the AST generator (though it may be used under the hood). I will pull down the example files and see if it resolves the issue around using it to get an AST of the tables with the proper start and end positions. I will add an update here around what the results are for the standalone files.

Thanks again for all the hard work you have put into this!

**Reply by pjkaufman** [reply]
You are correct that it is not reproducible. For some reason it must be some dependency convolution in `npm` that causes it in the project I was working in. Here is the example I used to verify that it was correct for the AST if anyone was currious:

`package.json`:
``` JSON
{
  "name": "example",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "remark-gfm": "^3.0.1",
    "remark": "^14.0.2",
    "unified": "^10.1.2"
  }
}
```

`example.js`:
``` JavaScript
import fs from "node:fs/promises";
import {visit} from 'unist-util-visit';
import {remark} from 'remark';
import remarkGfm from 'remark-gfm';

const file = await fs.readFile("example.md");
const fileContents = String(file);
const ast = remark().use(remarkGfm).parse(file);

console.log(ast)

visit(ast, 'table', (node) => {
  console.log(fileContents.substring(node.position.start.offset, node.position.end.offset) + '\nEND\n')

});
```

`example.md` is the same as the one linked to above.

The output looks as follows:
```
{
  type: 'root',
  children: [
    {
      type: 'heading',
      depth: 1,
      children: [Array],
      position: [Object]
    },
    {
      type: 'table',
      align: [Array],
      children: [Array],
      position: [Object]
    },
    {
      type: 'heading',
      depth: 1,
      children: [Array],
      position: [Object]
    },
    {
      type: 'table',
      align: [Array],
      children: [Array],
      position: [Object]
    }
  ],
  position: {
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 13, column: 1, offset: 373 }
  }
}
| Column 1 | Column 2 | Column 3 |
|----------|----------|----------|
| foo      | bar      | blob     |
| baz      | qux      | trust    |
| quux     | quuz     | glob     |
END

| Column 1 | Column 2 | Column 3 |
| -------- | -------- | -------- |
bar | baz | baq
foo | bar | bun
END
```

This is exactly as expected which leads me to believe that the issue is with how the packages are getting added together for me in the project. Thanks for all the help in getting me to a point where I could see that it was not actually an issue. I will see if I cannot resolve the issue of the dependency issues for me locally. But using the underlying packages directly does the trick for me.

Keep up all the great work!

**Reply by pjkaufman** [reply]
It seems that swapping my dependencies to expressly include unified has fixed it in the local project as well. I am not sure what happened. But whatever the issue was. It was on my end. Have a great weekend!
